File: webapp/src/Websocket/types/Rpc.ts
import {Subscribe, Unsubscribe} from "./Subscription";
import {ExecuteMatcherRequest} from "./transactionMatcher";
import {UserStateV2} from "./UserStateV2";
import {UnbookedBankTransactionMatcher} from "./unbookedTransactions";
import {Booking} from "./bookings";
import {AccountDto} from "./accounts";
import {BankAccount} from "./bankaccount";
import {User} from "./User";
import {UpdateBudget} from "./budget";

export type RequestType = "subscribe"
    | "unsubscribe"
    | 'importBankTransactions'
    | 'setUserStateV2'
    | 'deleteAllUnbookedTransactions'
    | 'deleteUnbookedTransaction'
    | 'addOrReplaceUnbookedTransactionMatcher'
    | 'removeUnbookedTransactionMatcher'
    | 'executeMatcherUnbookedTransactionMatcher'
    | 'createOrUpdateBooking'
    | 'deleteBooking'
    | 'updateChecked'
    | 'mergeAccount'
    | 'createOrUpdateAccount'
    | 'deleteBankAccount'
    | 'createOrUpdateBankAccount'
    | 'closeNextMonth'
    | 'reopenPreviousMonth'
    // admin commands:
    | 'addOrReplaceUser'
    | 'deleteUser'
    | 'createNewBackup'
    | 'restoreBackup'
    | 'dropBackup'
    | 'updateBudgetRulesForAccount'
    ;

export type RpcRequest = {
    type: RequestType;
    subscribe?: Subscribe;
    unsubscribe?: Unsubscribe;

    accountId?: string;
    deleteUnbookedBankTransactionId?: number;
    mergeTargetAccountId?: string;
    importBankTransactionsRequest?: ImportBankTransactionsRequest;
    executeMatcherRequest?: ExecuteMatcherRequest;
    deleteMatcherId?: string;
    userStateV2?: UserStateV2;
    unbookedBankTransactionMatcher?: UnbookedBankTransactionMatcher;
    removeUnbookedTransactionMatcherId?: string;
    bookingId?: number;
    bookingEntryId?: number;
    bookingEntryChecked?: boolean;
    createOrUpdateBooking?: Booking;
    createOrUpdateAccount?: AccountDto;
    bankAccount?: BankAccount;
    user?: User;
    deleteUserId?: string;
    backupName?: string;
    updateBudget?: UpdateBudget;
}

export type RpcResponse = {
    subscriptionCreated?: boolean;
    subscriptionRemoved?: boolean;

    importBankTransactionsResult?: ImportBankTransactionsResult;
    editedBookingId?: number;
    error?: string;
}

export type ImportBankTransactionsRequest = {
    accountId: string;
    filename: string;
    dataBase64: string;
    duplicationHandlerType: DuplicationHandlerType;
}

export type ImportBankTransactionsResult = {
    imported: number;
    skipped: number;
}

export type DuplicationHandlerType = 'sameDateAndAmount' | 'sameDateAmountAndDescription'

File: src/main/kotlin/com/dehnes/accounting/api/dtos/Rpc.kt
package com.dehnes.accounting.api.dtos

import com.dehnes.accounting.bank.importers.ImportResult
import com.dehnes.accounting.database.AccountDto
import com.dehnes.accounting.database.BankAccount
import com.dehnes.accounting.database.Booking
import com.dehnes.accounting.database.UnbookedBankTransactionMatcher


enum class RequestType {
    subscribe,
    unsubscribe,

    importBankTransactions,

    setUserStateV2,

    deleteAllUnbookedTransactions,
    deleteUnbookedTransaction,
    addOrReplaceUnbookedTransactionMatcher,
    removeUnbookedTransactionMatcher,

    executeMatcherUnbookedTransactionMatcher,

    createOrUpdateBooking,
    deleteBooking,
    updateChecked,

    mergeAccount,
    createOrUpdateAccount,

    deleteBankAccount,
    createOrUpdateBankAccount,

    closeNextMonth,
    reopenPreviousMonth,

    // admin commands
    addOrReplaceUser,
    deleteUser,
    createNewBackup,
    restoreBackup,
    dropBackup,
}

data class RpcRequest(
    val type: RequestType,
    val subscribe: Subscribe?,
    val unsubscribe: Unsubscribe?,
    val accountId: String?,
    val deleteUnbookedBankTransactionId: Long?,
    val mergeTargetAccountId: String?,
    val importBankTransactionsRequest: ImportBankTransactionsRequest?,
    val executeMatcherRequest: ExecuteMatcherRequest?,
    val deleteMatcherId: String?,
    val userStateV2: UserStateV2?,
    val unbookedBankTransactionMatcher: UnbookedBankTransactionMatcher?,
    val removeUnbookedTransactionMatcherId: String?,
    val bookingId: Long?,
    val bookingEntryId: Long?,
    val bookingEntryChecked: Boolean?,
    val createOrUpdateBooking: Booking?,
    val createOrUpdateAccount: AccountDto?,
    val bankAccount: BankAccount?,
    val user: User?,
    val deleteUserId: String?,
    val backupName: String?,
)

data class RpcResponse(
    val subscriptionCreated: Boolean? = null,
    val subscriptionRemoved: Boolean? = null,

    val importBankTransactionsResult: ImportResult? = null,
    val editedBookingId: Long? = null,
    val error: String? = null,
)

data class ImportBankTransactionsRequest(
    val accountId: String,
    val filename: String,
    val dataBase64: String,
)

